File: crates/tui/companion/wire.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type { Readable, Writable } from 'node:stream';
import { TextDecoder } from 'node:util';

// Source-prototype compatibility identity, NOT a released artifact hash.
export const BUILD_ID = 'maka-tui-m0-ipc-1';
export const MAX_FRAME_BYTES = 16 * 1024;
export const MAX_ECHO_BYTES = 4 * 1024;
const decoder = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true });

export type Packet =
  | { type: 'hello' | 'ready'; build_id: string }
  | { type: 'echo' | 'echoed'; request_id: number; text: string }
  | { type: 'shutdown' | 'bye' };

function record(value: unknown): asserts value is Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('IPC payload must be an object');
  }
}

function keys(value: Record<string, unknown>, allowed: string[]): void {
  if (
    Object.keys(value).length !== allowed.length ||
    allowed.some((key) => !Object.hasOwn(value, key))
  ) {
    throw new Error('IPC payload has missing or unknown fields');
  }
}

/** Validate a shallow, bounded schema. No arbitrary operation forwarding exists. */
export function validate(value: unknown): Packet {
  record(value);
  switch (value.type) {
    case 'hello':
    case 'ready':
      keys(value, ['type', 'build_id']);
      if (
        typeof value.build_id !== 'string' ||
        value.build_id.length > 128 ||
        !/^[\x20-\x7e]+$/.test(value.build_id)
      ) {
        throw new Error('Invalid IPC build identity');
      }
      return { type: value.type, build_id: value.build_id };
    case 'echo':
    case 'echoed':
      keys(value, ['type', 'request_id', 'text']);
      if (
        !Number.isInteger(value.request_id) ||
        (value.request_id as number) < 1 ||
        (value.request_id as number) > 0xffff_ffff
      ) {
        throw new Error('Invalid IPC request identity');
      }
      if (
        typeof value.text !== 'string' ||
        !value.text.isWellFormed() ||
        Buffer.byteLength(value.text) > MAX_ECHO_BYTES
      ) {
        throw new Error('IPC echo exceeds its text budget or has invalid Unicode');
      }
      return { type: value.type, request_id: value.request_id as number, text: value.text };
    case 'shutdown':
    case 'bye':
      keys(value, ['type']);
      return { type: value.type };
    default:
      throw new Error('Unknown IPC packet type');
  }
}

export function encode(value: Packet): Buffer {
  const payload = Buffer.from(JSON.stringify(validate(value)), 'utf8');
  if (payload.length === 0 || payload.length > MAX_FRAME_BYTES)
    throw new Error('IPC frame exceeds budget');
  const header = Buffer.alloc(8);
  header.write('MKUI', 0, 'ascii');
  header.writeUInt16BE(1, 4);
  header.writeUInt16BE(payload.length, 6);
  return Buffer.concat([header, payload]);
}

/** At most one 16 KiB body is assembled, including when the pipe is fragmented. */
export async function* packets(input: Readable): AsyncGenerator<Packet> {
  const header = Buffer.alloc(8);
  let headerBytes = 0;
  let body: Buffer | undefined;
  let bodyBytes = 0;
  for await (const chunk of input) {
    if (!Buffer.isBuffer(chunk)) throw new Error('IPC stream must contain bytes');
    let offset = 0;
    while (offset < chunk.length) {
      if (body === undefined) {
        const count = Math.min(8 - headerBytes, chunk.length - offset);
        chunk.copy(header, headerBytes, offset, offset + count);
        headerBytes += count;
        offset += count;
        if (headerBytes !== 8) continue;
        const length = header.readUInt16BE(6);
        if (!header.subarray(0, 4).equals(Buffer.from('MKUI')) || header.readUInt16BE(4) !== 1) {
          throw new Error('IPC magic or wire version mismatch');
        }
        if (length === 0 || length > MAX_FRAME_BYTES) throw new Error('IPC frame exceeds budget');
        body = Buffer.alloc(length);
      }
      const count = Math.min(body.length - bodyBytes, chunk.length - offset);
      chunk.copy(body, bodyBytes, offset, offset + count);
      bodyBytes += count;
      offset += count;
      if (bodyBytes === body.length) {
        const packet = validate(JSON.parse(decoder.decode(body)));
        body = undefined;
        headerBytes = 0;
        bodyBytes = 0;
        yield packet;
      }
    }
  }
  if (headerBytes !== 0 || body !== undefined) throw new Error('Truncated IPC frame');
}

/** Sequential writes bound buffering; a stalled peer cannot retain a write forever. */
export async function writePacket(output: Writable, packet: Packet): Promise<void> {
  const bytes = encode(packet);
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    await Promise.race([
      new Promise<void>((resolve, reject) =>
        output.write(bytes, (error) => (error ? reject(error) : resolve())),
      ),
      new Promise<never>((_, reject) => {
        timer = setTimeout(() => reject(new Error('IPC write timeout')), 2_000);
      }),
    ]);
  } finally {
    clearTimeout(timer);
  }
}

File: crates/tui/src/wire.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

//! Bounded, versioned frames for the private M0 Companion experiment.

use std::io::{self, Write};

use serde::{Deserialize, Serialize};
use thiserror::Error;
use tokio::io::{AsyncRead, AsyncReadExt};

pub(super) const MAX_FRAME_BYTES: usize = 16 * 1024;
pub(super) const MAX_ECHO_BYTES: usize = 4096;
// This pairs the two source prototypes; it is not a released artifact hash.
pub(super) const BUILD_ID: &str = "maka-tui-m0-ipc-1";
const MAGIC: [u8; 4] = *b"MKUI";
const VERSION: u16 = 1;
const HEADER_BYTES: usize = 8;
const MAX_BUILD_ID_BYTES: usize = 128;

#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub(super) enum Message {
    Hello { build_id: String },
    Ready { build_id: String },
    Echo { request_id: u32, text: String },
    Echoed { request_id: u32, text: String },
    Shutdown {},
    Bye {},
}

#[derive(Debug, Error)]
pub(super) enum WireError {
    #[error("Companion pipe I/O failed: {0}")]
    Io(#[from] io::Error),
    #[error("invalid Companion frame magic")]
    Magic,
    #[error("unsupported Companion protocol version")]
    Version,
    #[error("Companion frame exceeds its byte budget or is empty")]
    FrameLength,
    // Do not render a serde diagnostic containing untrusted peer text in the TTY.
    #[error("invalid Companion JSON message")]
    Json(#[source] serde_json::Error),
    #[error("invalid Companion build identifier")]
    BuildId,
    #[error("Companion request identifiers must be nonzero")]
    RequestId,
    #[error("Companion echo exceeds its byte budget")]
    EchoLength,
}

impl Message {
    fn validate(&self) -> Result<(), WireError> {
        match self {
            Self::Hello { build_id } | Self::Ready { build_id }
                if build_id.is_empty() || build_id.len() > MAX_BUILD_ID_BYTES =>
            {
                Err(WireError::BuildId)
            }
            Self::Echo { request_id: 0, .. } | Self::Echoed { request_id: 0, .. } => {
                Err(WireError::RequestId)
            }
            Self::Echo { text, .. } | Self::Echoed { text, .. } if text.len() > MAX_ECHO_BYTES => {
                Err(WireError::EchoLength)
            }
            _ => Ok(()),
        }
    }
}

fn payload_length(header: [u8; HEADER_BYTES]) -> Result<usize, WireError> {
    if header[..4] != MAGIC {
        return Err(WireError::Magic);
    }
    if u16::from_be_bytes([header[4], header[5]]) != VERSION {
        return Err(WireError::Version);
    }
    let length = usize::from(u16::from_be_bytes([header[6], header[7]]));
    if length == 0 || length > MAX_FRAME_BYTES {
        return Err(WireError::FrameLength);
    }
    Ok(length)
}

fn decode_payload(payload: &[u8]) -> Result<Message, WireError> {
    let message: Message = serde_json::from_slice(payload).map_err(WireError::Json)?;
    message.validate()?;
    Ok(message)
}

/// Read one frame. Keep this future alive until completion or discard the pipe.
pub(super) async fn read(reader: &mut (impl AsyncRead + Unpin)) -> Result<Message, WireError> {
    let mut header = [0; HEADER_BYTES];
    reader.read_exact(&mut header).await?;
    // No payload allocation occurs until magic, version, and size are admitted.
    let length = payload_length(header)?;
    let mut payload = vec![0; length];
    reader.read_exact(&mut payload).await?;
    decode_payload(&payload)
}

/// Encode within the same bound enforced by the reader, including JSON escaping.
pub(super) fn encode(message: &Message) -> Result<Vec<u8>, WireError> {
    message.validate()?;
    let mut payload = BoundedPayload(Vec::new());
    if let Err(error) = serde_json::to_writer(&mut payload, message) {
        return if error.is_io() {
            Err(WireError::FrameLength)
        } else {
            Err(WireError::Json(error))
        };
    }
    let length = u16::try_from(payload.0.len()).map_err(|_| WireError::FrameLength)?;
    let mut frame = Vec::with_capacity(HEADER_BYTES + payload.0.len());
    frame.extend_from_slice(&MAGIC);
    frame.extend_from_slice(&VERSION.to_be_bytes());
    frame.extend_from_slice(&length.to_be_bytes());
    frame.extend_from_slice(&payload.0);
    Ok(frame)
}

struct BoundedPayload(Vec<u8>);

impl Write for BoundedPayload {
    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
        if bytes.len() > MAX_FRAME_BYTES.saturating_sub(self.0.len()) {
            return Err(io::Error::other("frame byte budget exceeded"));
        }
        self.0.extend_from_slice(bytes);
        Ok(bytes.len())
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn read_bytes(bytes: &[u8]) -> Result<Message, WireError> {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .build()
            .unwrap();
        let mut reader = bytes;
        runtime.block_on(read(&mut reader))
    }

    fn raw_frame(payload: &[u8]) -> Vec<u8> {
        let mut frame = Vec::from(MAGIC);
        frame.extend_from_slice(&VERSION.to_be_bytes());
        frame.extend_from_slice(&u16::try_from(payload.len()).unwrap().to_be_bytes());
        frame.extend_from_slice(payload);
        frame
    }

    #[test]
    fn frame_round_trips_multibyte_text_without_normalization() {
        let message = Message::Echo {
            request_id: 7,
            text: "中文 e\u{301} 👩‍💻\n\t".to_owned(),
        };
        assert_eq!(read_bytes(&encode(&message).unwrap()).unwrap(), message);
    }

    #[test]
    fn reader_rejects_oversized_header_without_reading_a_payload() {
        let mut frame = Vec::from(MAGIC);
        frame.extend_from_slice(&VERSION.to_be_bytes());
        frame.extend_from_slice(&u16::MAX.to_be_bytes());
        assert!(matches!(read_bytes(&frame), Err(WireError::FrameLength)));
    }

    #[test]
    fn reader_rejects_empty_payload_before_json_decoding() {
        assert!(matches!(
            read_bytes(&raw_frame(b"")),
            Err(WireError::FrameLength)
        ));
    }

    #[test]
    fn reader_rejects_invalid_magic() {
        let mut frame = raw_frame(br#"{"type":"bye"}"#);
        frame[0] = b'X';
        assert!(matches!(read_bytes(&frame), Err(WireError::Magic)));
    }

    #[test]
    fn reader_rejects_unknown_version() {
        let mut frame = raw_frame(br#"{"type":"bye"}"#);
        frame[5] = 2;
        assert!(matches!(read_bytes(&frame), Err(WireError::Version)));
    }

    #[test]
    fn reader_reports_truncated_header_as_unexpected_eof() {
        assert!(
            matches!(read_bytes(b"MKU"), Err(WireError::Io(error)) if error.kind() == io::ErrorKind::UnexpectedEof)
        );
    }

    #[test]
    fn reader_reports_truncated_payload_as_unexpected_eof() {
        let mut frame = raw_frame(br#"{"type":"bye"}"#);
        frame.pop();
        assert!(
            matches!(read_bytes(&frame), Err(WireError::Io(error)) if error.kind() == io::ErrorKind::UnexpectedEof)
        );
    }

    #[test]
    fn reader_rejects_non_utf8_payload() {
        assert!(matches!(
            read_bytes(&raw_frame(b"\xff")),
            Err(WireError::Json(_))
        ));
    }

    #[test]
    fn reader_rejects_unknown_fields_on_empty_variant() {
        assert!(matches!(
            read_bytes(&raw_frame(br#"{"type":"bye","extra":true}"#)),
            Err(WireError::Json(_))
        ));
    }

    #[test]
    fn reader_rejects_unknown_fields_on_data_variant() {
        assert!(matches!(
            read_bytes(&raw_frame(
                br#"{"type":"ready","build_id":"x","extra":true}"#
            )),
            Err(WireError::Json(_))
        ));
    }

    #[test]
    fn reader_rejects_unknown_message_type() {
        assert!(matches!(
            read_bytes(&raw_frame(br#"{"type":"execute"}"#)),
            Err(WireError::Json(_))
        ));
    }

    #[test]
    fn reader_rejects_duplicate_required_fields() {
        assert!(matches!(
            read_bytes(&raw_frame(
                br#"{"type":"ready","build_id":"x","build_id":"y"}"#
            )),
            Err(WireError::Json(_))
        ));
    }

    #[test]
    fn reader_admits_exactly_the_payload_budget() {
        let mut payload = Vec::from(br#"{"type":"bye"}"#);
        payload.resize(MAX_FRAME_BYTES, b' ');
        assert_eq!(read_bytes(&raw_frame(&payload)).unwrap(), Message::Bye {});
    }

    #[test]
    fn one_read_does_not_consume_the_following_frame() {
        let first = Message::Ready {
            build_id: BUILD_ID.to_owned(),
        };
        let mut bytes = encode(&first).unwrap();
        bytes.extend_from_slice(&encode(&Message::Bye {}).unwrap());
        let runtime = tokio::runtime::Builder::new_current_thread()
            .build()
            .unwrap();
        let mut reader = bytes.as_slice();
        runtime.block_on(read(&mut reader)).unwrap();
        assert_eq!(
            runtime.block_on(read(&mut reader)).unwrap(),
            Message::Bye {}
        );
    }

    #[test]
    fn reader_rejects_zero_request_identifier() {
        assert!(matches!(
            read_bytes(&raw_frame(br#"{"type":"echoed","request_id":0,"text":""}"#)),
            Err(WireError::RequestId)
        ));
    }

    #[test]
    fn echo_budget_counts_utf8_bytes() {
        let message = Message::Echo {
            request_id: 1,
            text: "中".repeat(MAX_ECHO_BYTES / 3 + 1),
        };
        assert!(matches!(encode(&message), Err(WireError::EchoLength)));
    }

    #[test]
    fn encoder_bounds_json_escape_expansion() {
        let message = Message::Echo {
            request_id: 1,
            text: "\0".repeat(MAX_ECHO_BYTES),
        };
        assert!(matches!(encode(&message), Err(WireError::FrameLength)));
    }

    #[test]
    fn invalid_json_diagnostic_does_not_echo_peer_control_sequences() {
        let error = read_bytes(&raw_frame(br#"{"type":"\u001b[31m"}"#)).unwrap_err();
        assert_eq!(error.to_string(), "invalid Companion JSON message");
    }
}
